lib/doorkeeper/models/concerns/hashable.rb

```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+module Doorkeeper
+  module Models
+    ##
+    # Hashable finder to provide lookups for input plaintext values which are
+    # mapped to their hashes before lookup.
+    module Hashable
+      extend ActiveSupport::Concern
+
+      module ClassMethods
+        # Returns an instance of the Doorkeeper::AccessToken with
+        # specific token value.
+        #
+        # @param attr [Symbol]
+        #   The token attribute we're looking with.
+        #
+        # @param token [#to_s]
+        #   token value (any object that responds to `#to_s`)
+        #
+        # @return [Doorkeeper::AccessToken, nil] AccessToken object or nil
+        #   if there is no record with such token
+        #
+        def find_by_optional_hash(attr, token)
```

Comment:
I think we can hide from the end-user our internal API and call this method something like `find_by_token` and do all the magic (use plain token or encrypted one if encryption activated) in it

Reply:
There is already `by_token` in the AccessGrant and AccessToken mixins that are meant as the public interface. These are user facing, so my suggestion would simply be to mark this method `:private` to make it clear it is not a public interface.

Reply:
It's a good idea to mark it as a private (internal) API, but it's opened for developers so anyone could easily use it.. I'll reconsider it in a meanwhile

Also I would still think about the naming..

Reply:
I believe `find_by_hashed_or_plain_token`, while a tad long, describes exactly what it does. It's identical to what the helper does. If token hashing is active, it will look for a hashed token, or fall back to looking up a plaintext token.

Reply:
It's great :+1: